server/src/route/cluster_based.rs

```diff
@@ -19,8 +26,58 @@ impl ClusterBasedRouter {
     }
 }
 
+// Parse the raw endpoint which should be the form: <domain_name>:<port>
+//
+// Returns `None` if fail to parse.
+fn try_parse_endpoint(raw: &str) -> Option<Endpoint> {
+    let (domain, raw_port) = raw.split_once(':')?;
+    let port: u16 = raw_port.parse().map(Some).unwrap_or(None)?;
+
+    let mut endpoint = Endpoint::default();
+    endpoint.set_ip(domain.to_string());
+    endpoint.set_port(port as u32);
+    Some(endpoint)
+}
+
+#[async_trait]
 impl Router for ClusterBasedRouter {
-    fn route(&self, _schema: &str, _req: RouteRequest) -> Result<Vec<Route>> {
-        todo!();
+    async fn route(&self, schema: &str, mut req: RouteRequest) -> Result<Vec<Route>> {
+        let route_tables_req = RouteTablesRequest {
+            schema_name: schema.to_string(),
+            table_names: req.take_metrics().into(),
+        };
+        let route_resp = self
+            .cluster
+            .route_tables(&route_tables_req)
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(ErrWithCause {
+                code: StatusCode::InternalError,
+                msg: "Fail to route tables by cluster",
+            })?;
+
+        let mut routes = Vec::with_capacity(route_resp.entries.len());
+
+        // Now we pick up the nodes who own the leader shard for the route response.
```

Comment:
Is leader only here a hardcode? Maybe we need to add one field in `RouteRequest` to let the client specify what role it wants. Like `LeaderOnly`, `InSync`, `NoSync`, `DontCare` etc.

Reply:
It's true, but for now the protocol hasn't supported such things yet. Let's add the information about roles afterwards when the protocol upgraded.